### PR TITLE
Improve derive macro completion

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
* Adds completions for standard derive macros (considering their dependencies on each other, so we don't get compile errors)
* Adds completions for custom derive macros that are in scope, if the proc macro feature is enabled in the settings
* Separates macro completion from other completions to avoid incorrect completion propositions

### CODE REVIEW

File: crates/ra_hir/src/code_model.rs [hunk 1]
```diff
@@ -762,13 +762,12 @@ impl MacroDef {
...
     /// Indicate it is a proc-macro
     pub fn is_proc_macro(&self) -> bool {
-        match self.id.kind {
-            hir_expand::MacroDefKind::Declarative => false,
-            hir_expand::MacroDefKind::BuiltIn(_) => false,
-            hir_expand::MacroDefKind::BuiltInDerive(_) => false,
-            hir_expand::MacroDefKind::BuiltInEager(_) => false,
-            hir_expand::MacroDefKind::CustomDerive(_) => true,
-        }
+        matches!(self.id.kind, MacroDefKind::CustomDerive(_))
```
Comment (edwin0cheng) [inline crates/ra_hir/src/code_model.rs:765]: ❤️

**Comment by SomeoneToIgnore** [comment]
I have no clue how to test the custom derive macro case in this module, hence added no tests for that.

File: crates/ra_ide/src/completion/complete_attribute.rs [hunk 2]
```diff
@@ -126,6 +129,92 @@ const ATTRIBUTES: &[AttrCompletion] = &[
...
+    }
+}
+
+fn parse_derive_input(derive_input: ast::TokenTree) -> Result<FxHashSet<String>, ()> {
+    match (derive_input.left_delimiter_token(), derive_input.right_delimiter_token()) {
+        (Some(left_paren), Some(right_paren))
+            if left_paren.kind() == SyntaxKind::L_PAREN
+                && right_paren.kind() == SyntaxKind::R_PAREN =>
+        {
+            Ok(derive_input
```
Comment (edwin0cheng) [inline crates/ra_ide/src/completion/complete_attribute.rs:172]: The following code might be not parse properly:

```rust
#[derive(serde::Serialize)]
```

Comment (SomeoneToIgnore) [reply crates/ra_ide/src/completion/complete_attribute.rs:172]: Indeed, it gives `[` and `]` as left and right delimeter tokens.
I'll com eup with some code that parses that, yet I wonder if we can somehow parse that better than with this ad-hoc method.

**Comment by edwin0cheng** [comment]
> I have no clue how to test the custom derive macro case in this module, hence added no tests for that.

Yeah, maybe we could add a new meta in [fixture][1] in the future. 

[1]: https://github.com/rust-analyzer/rust-analyzer/blob/89e1f97515c36ab97bd378d972cabec0feb6d77e/crates/ra_db/src/fixture.rs#L204

**Comment by edwin0cheng** [comment]
LGTM, except the parsing problem.

File: crates/ra_ide/src/completion/complete_attribute.rs (hunk 2 above)
Comment (SomeoneToIgnore) [reply crates/ra_ide/src/completion/complete_attribute.rs]: Fixed.

**Comment by edwin0cheng** [comment]
Thanks !

bors r+

**Pull request merged**